File: apps/api/src/files.ts
// File-input validation for POST /v1/execute (FILES-06/07).
//
// Mirrors the Go worker's runner.SanitizeWorkspacePath so the API rejects bad
// paths up-front with a clear 400, but the worker remains the source of truth
// (it re-sanitizes regardless — host-escape-only threat model, never trust the
// path). Also computes the total DECODED byte size so an over-cap request is
// rejected with 413 before anything is enqueued.

import type { FileInput } from "@teovilla/code-runner-contract";

export type FileValidationError =
  | { kind: "path"; name: string; message: string }
  | { kind: "base64"; name: string; message: string }
  | { kind: "ref"; name: string; message: string };

export interface FileValidationResult {
  totalBytes: number;
  error?: FileValidationError;
}

/**
 * Sanitize a wire file name the same way the worker does: reject empty,
 * absolute, and any ".." traversal segment; preserve legitimate subdirs.
 * Returns the safe relative path or an error message. Wire paths are always
 * forward-slash.
 */
export function sanitizeWorkspacePath(
  name: string,
): { ok: true; rel: string } | { ok: false; message: string } {
  if (name === "") return { ok: false, message: "file name is empty" };
  // Absolute (POSIX or any leading slash).
  if (name.startsWith("/")) {
    return {
      ok: false,
      message: `file name "${name}" is absolute; only relative workspace paths are allowed`,
    };
  }
  const segments = name.split("/");
  for (const seg of segments) {
    if (seg === "..") {
      return {
        ok: false,
        message: `file name "${name}" contains a ".." traversal segment`,
      };
    }
  }
  // Collapse "." and empty segments while preserving order.
  const cleaned = segments.filter((s) => s !== "" && s !== ".").join("/");
  if (cleaned === "") {
    return {
      ok: false,
      message: `file name "${name}" does not resolve to a file inside the workspace`,
    };
  }
  return { ok: true, rel: cleaned };
}

// Strict base64 (no whitespace, valid padding). Node's Buffer.from is lenient,
// so we validate the shape explicitly before trusting a decode.
const BASE64_RE = /^[A-Za-z0-9+/]*={0,2}$/;

// A content-addressed blob ref: "sha256:" + 64 lowercase hex. Mirrors the wire
// schema's FileInput.ref pattern and the worker's validateFileInput.
const REF_RE = /^sha256:[a-f0-9]{64}$/;

/**
 * Validate every input file and sum the DECODED byte size. Returns the first
 * error encountered (path or base64). For utf8 files the decoded size is the
 * UTF-8 byte length of the content; for base64 it is the decoded byte length.
 */
export function validateFiles(files: FileInput[]): FileValidationResult {
  let totalBytes = 0;
  for (const f of files) {
    const path = sanitizeWorkspacePath(f.name);
    if (!path.ok) {
      return { totalBytes, error: { kind: "path", name: f.name, message: path.message } };
    }

    // content/ref XOR (BLOB, Phase 16). The generated FileInputSchema can't
    // express XOR (both fields are optional), so enforce it here AND in the
    // worker (never trust one side). A ref file carries no inline bytes — skip
    // base64/size accounting for it; the worker streams + sha256-verifies it.
    const hasContent = f.content !== undefined;
    const hasRef = f.ref !== undefined;
    if (hasContent && hasRef) {
      return {
        totalBytes,
        error: {
          kind: "ref",
          name: f.name,
          message: `file "${f.name}": exactly one of "content" or "ref" is allowed, not both`,
        },
      };
    }
    if (!hasContent && !hasRef) {
      return {
        totalBytes,
        error: {
          kind: "ref",
          name: f.name,
          message: `file "${f.name}": exactly one of "content" or "ref" is required`,
        },
      };
    }
    if (hasRef) {
      if (!REF_RE.test(f.ref!)) {
        return {
          totalBytes,
          error: {
            kind: "ref",
            name: f.name,
            message: `file "${f.name}": ref must be "sha256:<64 hex>", got "${f.ref}"`,
          },
        };
      }
      // Ref bytes are not counted toward MAX_FILES_BYTES (the inline-input cap):
      // CAS is exactly the path for large/shared files that exceed it.
      continue;
    }

    const encoding = f.encoding ?? "utf8";
    if (encoding === "base64") {
      const content = f.content ?? "";
      if (!BASE64_RE.test(content) || content.length % 4 !== 0) {
        return {
          totalBytes,
          error: {
            kind: "base64",
            name: f.name,
            message: `file "${f.name}": content is not valid base64`,
          },
        };
      }
      totalBytes += Buffer.from(content, "base64").length;
    } else {
      totalBytes += Buffer.byteLength(f.content ?? "", "utf8");
    }
  }
  return { totalBytes };
}

File: internal/runner/files.go
package runner

import (
	"encoding/base64"
	"fmt"
	"path"
	"strings"

	"github.com/teovillanueva/code-runner/packages/contract/gen/go/wire"
)

// decodeFileContent returns the raw bytes of a FileInput, interpreting
// f.Content according to f.Encoding.
//
//   - "" (absent) or "utf8" → the content string is taken verbatim as bytes
//     (back-compat: every existing text caller is unaffected).
//   - "base64"               → the content string is base64-decoded; a clear
//     error is returned on malformed input.
//
// The worker never trusts the API's validation, so this is the single decode
// path used by both the DockerSocketRunner and the ZygoteRunner relay.
func decodeFileContent(f wire.FileInput) ([]byte, error) {
	// content is now *string (a FileInput carries EXACTLY ONE of content/ref).
	// A nil Content here means the file was a blob ref that should have been
	// resolved into inline content BEFORE the runner saw it (the worker's
	// resolveBlobRefs runs first). Reaching here with nil Content is a
	// programming error, not user input — fail loud rather than write an empty
	// file. An empty inline string (Content set to "") is legitimate (a
	// zero-byte file) and handled by the "" pointer-deref below.
	if f.Content == nil {
		return nil, fmt.Errorf("file %q: has neither inline content nor a resolved blob (unresolved ref?)", f.Name)
	}
	content := *f.Content
	switch f.Encoding {
	case "", wire.FileInputEncodingUtf8:
		return []byte(content), nil
	case wire.FileInputEncodingBase64:
		raw, err := base64.StdEncoding.DecodeString(content)
		if err != nil {
			return nil, fmt.Errorf("file %q: invalid base64 content: %w", f.Name, err)
		}
		return raw, nil
	default:
		return nil, fmt.Errorf("file %q: unknown encoding %q (want utf8 or base64)", f.Name, f.Encoding)
	}
}

// SanitizeWorkspacePath converts a caller-supplied wire file name into a safe
// relative path anchored at the workspace root. Subdirectories are preserved
// (e.g. "data/input.csv"); absolute paths and any ".." traversal are rejected
// rather than silently collapsed, because the threat model is host-escape-only
// and a path that tries to escape is a signal, not noise.
//
// Wire paths are always forward-slash, so this uses the `path` package (not
// `filepath`) to stay platform-independent regardless of the worker's OS.
func SanitizeWorkspacePath(name string) (string, error) {
	if name == "" {
		return "", fmt.Errorf("file name is empty")
	}
	if path.IsAbs(name) {
		return "", fmt.Errorf("file name %q is absolute; only relative workspace paths are allowed", name)
	}
	// Reject any ".." segment outright rather than silently collapsing it.
	// path.Clean("/"+"../x") would fold to "/x" (inside the root), but the
	// threat model treats an attempted traversal as a signal, not noise: the
	// worker never trusts the path, so a "../" anywhere is an error.
	for _, seg := range strings.Split(name, "/") {
		if seg == ".." {
			return "", fmt.Errorf("file name %q contains a '..' traversal segment", name)
		}
	}
	// Clean folds redundant "." and "//" while preserving legitimate subdirs.
	clean := path.Clean("/" + name)
	rel := strings.TrimPrefix(clean, "/")
	if rel == "" || rel == "." {
		return "", fmt.Errorf("file name %q does not resolve to a file inside the workspace", name)
	}
	return rel, nil
}
